### PR TITLE
Workaround for NumPy builds that ship with a broken Dlpack deleter

### LINKED ISSUE #88082: Crash with `numpy >=1.23.4` and `torch 1.13.0` due to `DLpack` interaction
### 🐛 Describe the bug

We use PyTorch in our ML framework [Thinc](https://github.com/explosion/thinc), and we started noticing CI failures on both Windows and Linux after `torch 1.13.0` was released late last week. After isolating the delinquent test case, we were able to reproduce the crash locally with the following minimum reproducible example:

```python
import torch

x = torch.zeros(10)
data_np = x.detach().cpu().numpy()
rev = torch.utils.dlpack.from_dlpack(data_np)
x.data = rev
```

`stderr` output:
```
Fatal Python error: PyEval_SaveThread: the function must be called with the GIL held, but the GIL is released (the current Python thread state is NULL)
Python runtime state: initialized

Current thread 0x00007f8a356bd480 (most recent call first):
File "/home/user/regression-testing/torch_segfault.py", line 60 in <module>

Extension modules: numpy.core._multiarray_umath, numpy.core._multiarray_tests, numpy.linalg._umath_linalg, numpy.fft._pocketfft_internal, numpy.random._common, numpy.random.bit_generator, numpy.random._bounded_integers, numpy.random._mt19937, numpy.random.mtrand, numpy.random._philox, numpy.random._pcg64, numpy.random._sfc64, numpy.random._generator, torch._C, torch._C._fft, torch._C._linalg, torch._C._nested, torch._C._nn, torch._C._sparse, torch._C._special (total: 20)
[1]    8496 IOT instruction  python ~/regression-testing/torch_segfault.py
```

Attaching a debugger to the Python process yielded the following stack trace:
```
libc.so.6!__GI_raise(int sig) (/build/glibc-SzIz7B/glibc-2.31/sysdeps/unix/sysv/linux/raise.c:50)
libc.so.6!__GI_abort() (/build/glibc-SzIz7B/glibc-2.31/stdlib/abort.c:79)
fatal_error_exit(int status) (/usr/local/src/conda/python-3.10.6/Python/pylifecycle.c:2556)
fatal_error(int fd, int header, const char * prefix, const char * msg, int status) (/usr/local/src/conda/python-3.10.6/Python/pylifecycle.c:2667)
_Py_FatalErrorFunc(const char * func, const char * msg) (/usr/local/src/conda/python-3.10.6/Python/pylifecycle.c:2753)
_Py_FatalError_TstateNULL(const char * func) (/usr/local/src/conda/python-3.10.6/Python/ceval.c:271)
_Py_EnsureFuncTstateNotNULL() (/usr/local/src/conda/python-3.10.6/Include/internal/pycore_pystate.h:95)
PyEval_SaveThread() (/usr/local/src/conda/python-3.10.6/Python/ceval.c:529)
PyEval_SaveThread() (/usr/local/src/conda/python-3.10.6/Python/ceval.c:520)
libtorch_python.so!THPVariable_clear(THPVariable*) (Unknown Source:0)
libtorch_python.so!THPVariable_subclass_dealloc(_object*) (Unknown Source:0)
_multiarray_umath.cpython-310-x86_64-linux-gnu.so!array_dealloc (Unknown Source:0)
libc10.so!c10::deleteInefficientStdFunctionContext(void*) (Unknown Source:0)
libtorch_python.so!c10::StorageImpl::~StorageImpl() (Unknown Source:0)
libc10.so!c10::intrusive_ptr<c10::StorageImpl, c10::detail::intrusive_target_default_null_type<c10::StorageImpl> >::reset_() (Unknown Source:0)
libc10.so!c10::TensorImpl::~TensorImpl() (Unknown Source:0)
libc10.so!c10::TensorImpl::~TensorImpl() (Unknown Source:0)
libtorch_python.so!THPVariable_clear(THPVariable*) (Unknown Source:0)
libtorch_python.so!THPVariable_subclass_dealloc(_object*) (Unknown Source:0)
_Py_Dealloc(PyObject * op) (/usr/local/src/conda/python-3.10.6/Objects/object.c:2298)
_Py_DECREF(PyObject * op) (/usr/local/src/conda/python-3.10.6/Include/object.h:500)
_Py_XDECREF(PyObject * op) (/usr/local/src/conda/python-3.10.6/Include/object.h:567)
free_keys_object(PyDictKeysObject * keys) (/usr/local/src/conda/python-3.10.6/Objects/dictobject.c:628)
dictkeys_decref(PyDictKeysObject * dk) (/usr/local/src/conda/python-3.10.6/Objects/dictobject.c:337)
dict_dealloc(PyDictObject * mp) (/usr/local/src/conda/python-3.10.6/Objects/dictobject.c:2076)
odict_dealloc(PyODictObject * self) (/usr/local/src/conda/python-3.10.6/Objects/odictobject.c:1416)
_Py_Dealloc(PyObject * op) (/usr/local/src/conda/python-3.10.6/Objects/object.c:2298)
_Py_DECREF(PyObject * op) (/usr/local/src/conda/python-3.10.6/Include/object.h:500)
_Py_XDECREF(PyObject * op) (/usr/local/src/conda/python-3.10.6/Include/object.h:567)
dict_dealloc(PyDictObject * mp) (/usr/local/src/conda/python-3.10.6/Objects/dictobject.c:2068)
_Py_Dealloc(PyObject * op) (/usr/local/src/conda/python-3.10.6/Objects/object.c:2298)
_Py_DECREF(PyObject * op) (/usr/local/src/conda/python-3.10.6/Include/object.h:500)
subtype_dealloc(PyObject * self) (/usr/local/src/conda/python-3.10.6/Objects/typeobject.c:1438)
_Py_Dealloc(PyObject * op) (/usr/local/src/conda/python-3.10.6/Objects/object.c:2298)
_Py_DECREF(PyObject * op) (/usr/local/src/conda/python-3.10.6/Include/object.h:500)
_Py_XDECREF(PyObject * op) (/usr/local/src/conda/python-3.10.6/Include/object.h:567)
dict_dealloc(PyDictObject * mp) (/usr/local/src/conda/python-3.10.6/Objects/dictobject.c:2068)
_Py_Dealloc(PyObject * op) (/usr/local/src/conda/python-3.10.6/Objects/object.c:2298)
_Py_DECREF(PyObject * op) (/usr/local/src/conda/python-3.10.6/Include/object.h:500)
subtype_dealloc(PyObject * self) (/usr/local/src/conda/python-3.10.6/Objects/typeobject.c:1438)
_Py_Dealloc(PyObject * op) (/usr/local/src/conda/python-3.10.6/Objects/object.c:2298)
_Py_DECREF(PyObject * op) (/usr/local/src/conda/python-3.10.6/Include/object.h:500)
_Py_XDECREF(PyObject * op) (/usr/local/src/conda/python-3.10.6/Include/object.h:567)
list_dealloc(PyListObject * op) (/usr/local/src/conda/python-3.10.6/Objects/listobject.c:349)
_Py_DECREF(PyObject * op) (/usr/local/src/conda/python-3.10.6/Include/object.h:500)
clear_slots(PyTypeObject * type) (/usr/local/src/conda/python-3.10.6/Objects/typeobject.c:1267)
subtype_clear(PyObject * self) (/usr/local/src/conda/python-3.10.6/Objects/typeobject.c:1285)
delete_garbage(PyGC_Head * old, PyGC_Head * collectable, GCState * gcstate, PyThreadState * tstate) (/usr/local/src/conda/python-3.10.6/Modules/gcmodule.c:1017)
gc_collect_main(PyThreadState * tstate, int generation, Py_ssize_t * n_collected, Py_ssize_t * n_uncollectable, int nofail) (/usr/local/src/conda/python-3.10.6/Modules/gcmodule.c:1300)
_PyGC_CollectNoFail(PyThreadState * tstate) (/usr/local/src/conda/python-3.10.6/Modules/gcmodule.c:2123)
finalize_modules(PyThreadState * tstate) (/usr/local/src/conda/python-3.10.6/Python/pylifecycle.c:1528)
Py_FinalizeEx() (/usr/local/src/conda/python-3.10.6/Python/pylifecycle.c:1787)
Py_RunMain() (/usr/local/src/conda/python-3.10.6/Modules/main.c:672)
Py_BytesMain(int argc, char ** argv) (/usr/local/src/conda/python-3.10.6/Modules/main.c:1090)
libc.so.6!__libc_start_main(int (*)(int, char **, char **) main, int argc, char ** argv, int (*)(int, char **, char **) init, void (*)(void) fini, void (*)(void) rtld_fini, void * stack_end) (/build/glibc-SzIz7B/glibc-2.31/csu/libc-start.c:308)
_start() (/usr/local/src/conda/python-3.10.6/Python/getplatform.c:9)
```

### Versions
```
PyTorch version: 1.13.0+cpu
Is debug build: False
CUDA used to build PyTorch: Could not collect
ROCM used to build PyTorch: N/A

OS: KDE neon 5.26 (x86_64)
GCC version: (Ubuntu 11.3.0-1ubuntu1~22.04) 11.3.0
Clang version: Could not collect
CMake version: version 3.22.1
Libc version: glibc-2.35

Python version: 3.10.6 (main, Oct 24 2022, 16:07:47) [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-5.15.0-52-generic-x86_64-with-glibc2.35
Is CUDA available: False
CUDA runtime version: 11.3.109
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: GPU 0: NVIDIA GeForce RTX 3090
Nvidia driver version: 515.65.01
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.6.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.6.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.6.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.6.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.6.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.6.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.6.0
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] mypy==0.982
[pip3] mypy-extensions==0.4.3
[pip3] numpy==1.23.4
[pip3] torch==1.13.0+cpu
[conda] numpy                     1.23.4                   pypi_0    pypi
[conda] torch                     1.13.0+cpu               pypi_0    pypi
```

cc @ezyang @gchanan @zou3519 @albanD

### PR DESCRIPTION
**Pull request by shadeMe**
NumPy versions 1.22 and 1.23 (and their respective bugfix releases included) have a buggy implementation of the Dlpack deleter that doesn't account for no-GIL contexts. Since we now release the GIL when deallocating tensors in `THPVariable_clear`, this leads to a failure of internal consistency checks when freeing a Dlpack-backed tensor from NumPy.

This PR adds a check for the buggy NumPy versions and overrides the `DlManagedTensor` deleter to reacquire the GIL before deallocation.

### Rationale for this implementation
The version check was added to `tensor_numpy.h/cpp` as it seemed like a more logical location for it than creating a new translation unit. The overriding of the deleter was originally attempted by directly modifying `at::fromDlpack`, but the lack of a build dependency on the Python C API in A10 prevented that. So, I extended the A10 Dlpack API instead to additionally accept a custom deleter functor.

Fixes #88082

### CODE REVIEW

**Comment by shadeMe** [comment]
I'm in the process of getting my organization's approval for the CLA. Will hopefully happen over the next couple of days.

**Review by albanD — COMMENTED** [review]
That sounds good!
I wonder if there is a way for us to test this?

**Comment by shadeMe** [comment]
> I wonder if there is a way for us to test this?
Outside of creating a `venv` with one of the buggy NumPy versions, executing the repro-example and checking for a non-zero return code from `python`, I don't immediately see any straight-forward way to test this.

**Review by albanD — APPROVED** [review]
Sounds good to me.
I agree that getting CI for this is going to be more work than it's worth and locally making sure it works is enough.

Just waiting on the CLA (you can comment /easycla to trigger a recheck once you're done with it) and then we can merge!

**Comment by shadeMe** [comment]
Seems like this PR didn't make the 1.13.1 release. Is there another bugfix release planned for 1.13?

**Comment by albanD** [comment]
None that is planned right now AFAIK. cc @atalman ?

**Review by albanD — COMMENTED** [review]
This looks good but the windows failure is real.

**Comment by shadeMe** [comment]
> This looks good but the windows failure is real.

Ah, apologies - I overlooked that one. I've fixed it now.

**Review by albanD — APPROVED** [review]

**Comment by albanD** [comment]
Ho lint needs fixing!

**Comment by shadeMe** [comment]
Rebased on `master` and tested the changes locally. The other failed checks from earlier seem to be unrelated to the changes in the PR, as far as I can tell.

**Pull request closed without merging**
